### ISSUE TITLE
Suggestion: SharedProxyOptions.MessageHandler should be a Func that returns the MessageHandler

### ISSUE THREAD

**Issue by aspnet-hello** [aspnet/Home#2489]
_From @adamholdenyall on Thursday, November 2, 2017 2:08:28 PM_

I have spent a good chunk of time chasing down an issue where a long running request through the proxy deadlocked other short requests until it finished. I finally narrowed it down to the fact that it was using the same MessageHandler for every request, and HttpClientHandler only seems to support one request at a time (I am on an older version of Core fwiw). 

It may be best to make this option a function so that the user can create a new instance of the MessageHandler for each request to avoid deadlocks.

_Copied from original issue: aspnet/Proxy#72_

**Comment by aspnet-hello** [comment]
_From @mkosieradzki on Thursday, November 2, 2017 2:17:52 PM_

What is the value for HttpClientHandler.MaxConnectionsPerServer in your handler?

**Comment by aspnet-hello** [comment]
_From @adamholdenyall on Thursday, November 2, 2017 2:23:05 PM_

I am targeting net452, so as per [this API doc](https://msdn.microsoft.com/en-us/library/system.net.http.httpclienthandler.maxconnectionsperserver(v=vs.110).aspx) it looks like that property isn't available to me.

**Comment by aspnet-hello** [comment]
_From @mkosieradzki on Thursday, November 2, 2017 2:39:11 PM_

I don't believe this library is intended to target anything below net 461, but I might be wrong,

There was a lot of mess related to different HTTP client handler implementations throughout the time and I believe only implementations from the newer .NET Standard will work correctly.
My guess will be to use anything higher than System.Net.Http nuget package 4.3.1. (.net 45 support seems to be non-existent in this package). I think that .net 46 is the lowest that should work fine.

Maybe others have something more to say on that, but I think we can expect any kind of strange behavior on the earlier versions of System.Net.Http.

If you need net452 support you can provide your own HttpMessageHandler implementation that will pool multiple real HttpClientHandlers undearneath.
Basically you can achieve this suggested "func" behavior just by subclassing HttpMessageHandler.

However if you will be able to reproduce the starvation issue on the newer net framework versions I think it will definitely require more attention in the default implementation as it is far from acceptable.

**Comment by aspnet-hello** [comment]
_From @adamholdenyall on Thursday, November 2, 2017 2:45:13 PM_

net452 is pretty old at this point, but I'm sure there are plenty of people on 461 and 462 are just barely older than a year at this point, and they don't look to have MaxConnectionsPerServer either. Even if you are targeting a newer framework, depending on what the default of that property is, it may be worth bringing up this configuration in the README of this project as well as a workaround for older frameworks.

I have essentially done the workaround that you described, and I am avoiding the deadlock.

EDIT: It should also be made clear that I am still using the 0.2.0 NuGet package as its the newest available.

**Comment by aspnet-hello** [comment]
_From @mkosieradzki on Thursday, November 2, 2017 2:56:47 PM_

Yup, great point about documentation.

However if you install System.Net.Http 4.3.1 nuget package over .NET 4.6 project or newer you will get the OOB HttpClientHandler which has this property.

**Comment by aspnet-hello** [comment]
_From @adamholdenyall on Thursday, November 2, 2017 3:23:33 PM_

Interesting info, I tried that NuGet package in a new project and saw the property become available. Good to know. Might be work upping the target of my main project if it's not much effort to get that pooling.

On a separate, but related issue, all these issues came up with a realllly non-performant request hosing up my proxy. Not only did this cause the aforementioned deadlock, but some requests timed out and cancelled the thread before the proxy returned. I don't see a configurable timeout setting in this project or on the newer version of the HttpClientHandler. Am I missing something?

**Comment by aspnet-hello** [comment]
_From @mkosieradzki on Thursday, November 2, 2017 3:41:46 PM_

I think adding a configurable timeout is a nice idea. It is possible to achieve it, however you need to use the new dynamic routing APIs and cancellation tokens. I am not sure if the package with dynamic routing has been already released.

If you think it's a good idea to add a global timeout setting please create a separate issue for it.

However just to make it clear: I am not maintainer of the project, just one of the contributors.

To sum up: the follow-up of this thread should be blocking up support for earlier versions of .NET than 4.6. However it has been already blocked - see https://github.com/aspnet/Proxy/blob/dev/src/Microsoft.AspNetCore.Proxy/Microsoft.AspNetCore.Proxy.csproj - it requires .NET Standard 2.0 i.e. it requires .NET 4.6.1.

And now my personal recommendation: today I would recommend targeting at least .NET 4.6.1 which allows you to utilize the entire .NET Standard 2.0 (and also being consumed by .NET Standard 2.0 libraries). The next big milestone is the very new.NET 4.7.1 which has entire .NET Standard 2.0 built in.

**Comment by aspnet-hello** [comment]
_From @Tratcher on Friday, November 3, 2017 8:25:39 AM_

A new handler per request is a non-starter, it would destroy your performance and exhaust your available sockets. Really this is a question of properly configuring the handler so it does not cause bottlenecks. For net461 you set ServicePointManager.DefaultConnectionLimit to 50 or so.

The timeout is a good idea. Otherwise it uses HttpClient's default of 100 seconds. Compare to how we set this up in the auth components:
https://github.com/aspnet/Security/blob/bd8ecd02684d490ece96c484e1092386f3515dec/src/Microsoft.AspNetCore.Authentication/RemoteAuthenticationOptions.cs#L66-L78
You can set the timeout and/or the handler it creates the client for you. Or you set the client directly.
